### ISSUE TITLE
import style best practice?

### ISSUE THREAD

**Issue by dschult** [networkx/networkx#4277]
We have some code that takes great pains to import code one function at a time. Other code imports lots of functions from the same module, but lists each function that is used in the module. Other code just imports the package name and uses it when needed to call a function from that package.

```python
from numpy import min
from numpy import max
from numpy import abs

x = min(abs(max(y)))
```
VS
```python
from numpy import min, max, abs

x = min(abs(max(y)))
```
VS
```python
import numpy as np

x = np.min(np.abs(np.max(y)))
```

Recently, I was trying to figure out where a package (numpy) was being used in a module. It would have been quite easy if the 3rd style was used.... I just search for "np." in my editor.  With either of the first two styles I have to search for min and max and abs and those words appear much more frequently (in comments and in docs as well as code) and I have to check whether the function has been defined as a local variable inside the function before being called (unlikely with this example min/max/abs, but i worried about it anyway).

Between the first two styles, I prefer the second because the code is more compact. And we used to always do it that way, but about 6 years ago people started using the more verbose style and even changing imports to that style.

Is there a Style guide preference for imports that anyone knows about? Or even a blog post by people who've thought about it some?  I think we should work toward making ours more uniform.

**Comment by rossbar** [comment]
My vote would definitely be for number 3. Aside from the code-searching consideration you mentioned (a very good point), there's a strong motivation to keep namespaces to avoid collisions with builtins (this is especially important for numpy, as there are many functions with the same names as builtins). For example:

```python
>>> my_trivial_genexpr = (i for i in [5, 3, 1])
>>> max(my_trivial_genexpr)
5
>>> from numpy import max
>>> max(my_trivial_genexpr)
<generator object <genexpr> at 0x7fb8198bbba0>
```

**Comment by jarrodmillman** [comment]
+1 for option 3
```
import numpy as np

x = np.min(np.abs(np.max(y)))
```

**Comment by jarrodmillman** [comment]
I started looking at ``linalg/algebraicconnectivity.py`` and it needs a lot of work.  I will refactor this before 2.6 and will take care of the imports at the same time.  I'll submit a draft PR in the next week.

**Comment by dschult** [comment]
I looked at [the PEP8 guidelines for imports](https://www.python.org/dev/peps/pep-0008/#imports) and I learned that
1) absolute import paths are preferred over relative imports.
2) ```__all__``` should be defined **before** the import statements, and after the module doc_string.
3) It doesn't specify a preference for or against specifying which names to import as opposed to importing ```np``` into the namespace. Though it does say that if there is a chance of confusion to use ```import numpy as np``` and the specify the function with the package name: ```np.max```.   

So, let's aim for option 3

```
import numpy as np

x = np.min(np.abs(np.max(y)))
```

There is a place in the Contributor Guide where we talk about imports. I'll add this there...
